lib/msf/core/modules/external/python/metasploit/probe_scanner.py

```diff
@@ -0,0 +1,99 @@
+    return lambda args: start_scanner(payload, pattern, args, onmatch, connect_timeout=connect_timeout, read_timeout=read_timeout)
+
+
+def start_scanner(payload, pattern, args, onmatch, **timeouts):
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(run_scanner(payload, pattern, args, onmatch, **timeouts))
+
+
+async def run_scanner(payload, pattern, args, onmatch, **timeouts):
+    probes = [probe_host(host, int(args['rport']), payload, **timeouts) for host in args['rhosts']]
+    async for (target, res) in Scan(probes):
+        if isinstance(res, Exception):
+            module.log('{}:{} - Error connecting: {}'.format(*target, res), level='error')
+        elif res and re.search(pattern, res):
+            module.log('{}:{} - Matches'.format(*target), level='good')
+            module.log('{}:{} - Matches with: {}'.format(*target, res), level='debug')
+            onmatch(target, res)
+        else:
+            module.log('{}:{} - Does not match'.format(*target), level='info')
+            module.log('{}:{} - Does not match with: {}'.format(*target, res), level='debug')
+
+
+class Scan:
+    def __init__(self, runs):
+        self.queue = asyncio.queues.Queue()
+        self.total = len(runs)
+        self.done = 0
+
+        for r in runs:
+            f = asyncio.ensure_future(r)
+            args = r.cr_frame.f_locals
+            target = (args['host'], args['port'])
+            f.add_done_callback(functools.partial(self.__queue_result, target))
+
+
+    def __queue_result(self, target, f):
+        res = None
+
+        try:
+            res = f.result()
+        except Exception as e:
+            res = e
+
+        self.queue.put_nowait((target, res))
+
+
+    async def __aiter__(self):
+        return self
+
+
+    async def __anext__(self):
+        if self.done == self.total:
+            raise StopAsyncIteration
+
+        res = await self.queue.get()
+        self.done += 1
+        return res
+
+
+async def probe_host(host, port, payload, connect_timeout, read_timeout):
+    r = None
+    w = None
+    buf = bytearray()
+
+    async with timeout(connect_timeout):
+        r, w = await asyncio.open_connection(host, port)
+        remote = w.get_extra_info('peername')
+        if remote[0] == host:
+            module.log('{}:{} - Connected'.format(host, port), level='debug')
+        else:
+            module.log('{}({}):{} - Connected'.format(host, *remote), level='debug')
+        w.write(payload)
+        await w.drain()
+
+    try:
+        async with timeout(read_timeout):
+            while len(buf) < 4096:
+                data = await r.read(4096)
+                if data:
+                    module.log('{}:{} - Received {} bytes'.format(host, port, len(data)), level='debug')
+                    buf.extend(data)
+                else:
+                    break
+    except asyncio.TimeoutError:
+        if buf:
+            pass
+        else:
+            raise
+
+    w.close()
```

Comment:
What do you think about moving this line into a `finally` block to ensure it happens prefaced with the check `if w is not None:`?

Reply:
If `w` is `None` I don't see how we could get here. If it doesn't close on a few cases, it's not that bad a problem, as these are short-lived processes.

Reply:
It might be worth wrapping it in a `try` block, though. If we have a response, not being able to close the connection shouldn't be a problem.

Reply:
Regardless of it being short-lived, wouldn't it be a good idea to ensure we cleanup after ourselves?

Reply:
We do, by exiting the process at the end of a batch. Python will also automatically close garbage-collected sockets, and `close()`d sockets aren't immediately shutdown anyway. Adding all the extra defensiveness doesn't improve how well the code works and hinders readability.

Reply:
I realize this is simply a scanner, but we are creating examples that will be followed. Per [Python Socket Programming HOWTO](https://docs.python.org/3/howto/sockets.html#disconnecting) 
> Python takes the automatic shutdown a step further, and says that when a socket is garbage collected, it will automatically do a close if it’s needed. But relying on this is a very bad habit. If your socket just disappears without doing a close, the socket at the other end may hang indefinitely, thinking you’re just being slow. Please close your sockets when you’re done.